packages/server/modules/workspaces/helpers/sso.ts

```diff
@@ -0,0 +1,70 @@
+import { getEncryptionKeyPair } from '@/modules/automate/services/encryption'
+import { getFrontendOrigin, getServerOrigin } from '@/modules/shared/helpers/envHelper'
+import { buildDecryptor, buildEncryptor } from '@/modules/shared/utils/libsodium'
+import { Request } from 'express'
+
+/**
+ * Generate Speckle URL to redirect users to after they complete authorization
+ * with the given SSO provider.
+ */
+export const buildAuthRedirectUrl = (
+  workspaceSlug: string,
+  isValidationFlow: boolean
+): URL => {
+  const urlFragments = [`/api/v1/workspaces/${workspaceSlug}/sso/oidc/callback`]
+
+  if (isValidationFlow) {
+    urlFragments.push('?validate=true')
+  }
+
+  return new URL(urlFragments.join(''), getServerOrigin())
+}
+
+/**
+ * Generate Speckle URL to redirect users to after successfully completing the
+ * SSO authorization flow.
+ * @remarks Append params to this URL to preserve information about errors
+ */
+export const buildFinalizeUrl = (workspaceSlug: string): URL => {
+  const urlFragments = [`workspaces/${workspaceSlug}/sso`]
+
+  return new URL(urlFragments.join(''), getFrontendOrigin())
+}
+
+/**
+ * Generate Speckle URL to redirect users to after an error occurs during SSO.
+ */
+export const buildErrorUrl = (err: unknown, workspaceSlug: string) => {
+  const errorRedirectUrl = buildFinalizeUrl(workspaceSlug)
+  const errorMessage = err instanceof Error ? err.message : `Unknown error: ${err}`
+  errorRedirectUrl.searchParams.set('error', errorMessage)
+  return errorRedirectUrl.toString()
+}
+
+export const encryptorFactory = () => async (data: string) => {
+  const encryptionKeyPair = await getEncryptionKeyPair()
+  const encryptor = await buildEncryptor(encryptionKeyPair.publicKey)
+  const encryptedData = await encryptor.encrypt(data)
+
+  encryptor.dispose()
+
+  return encryptedData
+}
+
+export const decryptorFactory = () => async (data: string) => {
+  const encryptionKeyPair = await getEncryptionKeyPair()
+  const decryptor = await buildDecryptor(encryptionKeyPair)
+  const decryptedData = await decryptor.decrypt(data)
+
+  decryptor.dispose()
+
+  return decryptedData
+}
+
+export const parseCodeVerifier = async (req: Request<unknown>): Promise<string> => {
+  const encryptedCodeVerifier = req.session.codeVerifier
+  if (!encryptedCodeVerifier)
+    throw new Error('Cannot find verification token. Restart flow.')
```

Comment:
The bare Error types should be replaced with types inheriting from our `BaseError`, where the messages can be made static and the response status codes must be defined too